Non-viable test classloader when running a group of tests (Class or Package) in VSCode

### Describe the bug

In VSCode, when running a group of tests (all tests in a class or all tests in a package), the tests fail with: 

```
org.junit.jupiter.engine.execution.ConditionEvaluationException: Failed to evaluate condition [io.quarkus.test.junit.QuarkusTestExtension]: Non-viable test classloader, io.quarkus.test.junit.classloading.FacadeClassLoader@1bc715b8. Is this a re-run of a failing test?
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1602)
        at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
        at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.IllegalStateException: Non-viable test classloader, io.quarkus.test.junit.classloading.FacadeClassLoader@1bc715b8. Is this a re-run of a failing test?
        at io.quarkus.test.junit.AbstractJvmQuarkusTestExtension.evaluateExecutionCondition(AbstractJvmQuarkusTestExtension.java:169)
```

If I run a single test, it passes.

Opening this at the request of @holly-cummins, I'm running into this while working on https://github.com/commonhaus/automation. 

- Quarkus 3.22.3
- Language Support for Java(TM) by Red Hat (1.42.0)
- Test Runner for Java (0.43.1)


### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/AbstractJvmQuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/AbstractJvmQuarkusTestExtension.java
@@ -162,8 +162,16 @@ public class AbstractJvmQuarkusTestExtension extends AbstractQuarkusTestWithCont
         } catch (Exception | ServiceConfigurationError e) {
             boolean isEclipse = System.getProperty("sun.java.command") != null
                     && System.getProperty("sun.java.command").contains("JUnit5TestLoader");
-            if (isEclipse) {
-                // Tracked by https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/2257
+
+            // VS Code has the exact same java command and runner as Eclipse, but needs its own message
+            boolean isVSCode = isEclipse && System.getProperty("java.class.path").contains("vscode");
+
+            if (isVSCode) {
+                // Will need https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/2257 and a reconsume by VSCode
+                Log.error(
+                        "Could not read configuration while evaluating whether to run a test. This is a known issue when running tests in the VS Code IDE. To work around the problem, run individual test methods.");
+            } else if (isEclipse) {
+                // Tracked by https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/2257; fixed in Eclipse 4.37
                 Log.error(
                         "Could not read configuration while evaluating whether to run a test. This is a known issue when running tests in the Eclipse IDE. To work around the problem, edit the run configuration and add `-uniqueId [engine:junit-jupiter]/[class:"
                                 + context.getRequiredTestClass().getName()
@@ -177,8 +185,14 @@ public class AbstractJvmQuarkusTestExtension extends AbstractQuarkusTestWithCont
             Log.debug("Underlying exception: " + e);
             Log.debug("Thread Context Classloader: " + Thread.currentThread().getContextClassLoader());
             Log.debug("The class of the class we use for mapping is " + TestConfig.class.getClassLoader());
-            throw new IllegalStateException("Non-viable test classloader, " + Thread.currentThread().getContextClassLoader()
-                    + ". Is this a re-run of a failing test?");
+            String message = isVSCode
+                    ? "Could not execute test class because it was loaded with the wrong classloader by the VS Code test runner. Try running test methods individually instead."
+                    : isEclipse
+                            ? "Could not execute test class because it was loaded with the wrong classloader by the Eclipse test runner. Try running test methods individually, or edit the run configuration and add `-uniqueId [engine:junit-jupiter]/[class:"
+                                    + context.getRequiredTestClass().getName()
+                                    + "]` in the program arguments. "
+                            : "Internal error: Test class was loaded with an unexpected classloader or the thread context classloader was incorrect.";
+            throw new IllegalStateException(message, e);
         } finally {
             if (!isFlatClasspath) {
                 Thread.currentThread().setContextClassLoader(original);

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/launcher/CustomLauncherInterceptor.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/launcher/CustomLauncherInterceptor.java
@@ -67,6 +67,10 @@ public class CustomLauncherInterceptor implements LauncherDiscoveryListener, Lau
 
     @Override
     public void launcherDiscoveryStarted(LauncherDiscoveryRequest request) {
+        // If anything comes through this method for which there are non-null classloaders on the selectors, that will bypass our classloading
+        // To check that case, the code would be something like this. We could detect and warn early, and possibly even filter that test out, but that's not necessarily a better UX than failing later
+        // request.getSelectorsByType(ClassSelector.class).stream().map(ClassSelector::getClassLoader) ... and then check for non-emptiness on that field
+
         // Do not do any classloading dance for prod mode tests;
         if (!isProductionModeTests()) {
             initializeFacadeClassLoader();

